File: gtci/10-level-order-traversal/cpp/5-zigzag-traversal.cpp
#include <cassert>
#include <deque>
#include <iostream>
#include <queue>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
  vector<vector<int>> traverse(TreeNode *root) {
    if (root == nullptr)
      return {};

    vector<vector<int>> traversal;
    
    std::queue<TreeNode*> queue;
    queue.push(root);

    bool zigzag = false;

    while (!queue.empty()) {
      const size_t levelSize = queue.size();

      std::deque<int> level;
      for (size_t i = 0; i < levelSize; i++) {
        TreeNode* node = queue.front();
        queue.pop();

        if (zigzag)
          level.push_front(node->val);
        else
          level.push_back(node->val);
        
        if (node->left != nullptr)
          queue.push(node->left);
        
        if (node->right != nullptr)
          queue.push(node->right);
      }

      traversal.push_back(std::vector<int>(level.begin(), level.end()));
      zigzag = !zigzag;
    }
    
    return traversal;
  }
};

int main() {
            Solution sol = Solution();

    TreeNode* root1 = new TreeNode(4);
    root1->left = new TreeNode(5);
    root1->right = new TreeNode(10);
    root1->left->left = new TreeNode(5);
    root1->left->right = new TreeNode(7);
    std::vector<vector<int>> expected1 = {{4}, {10, 5}, {5, 7}};
    assert(sol.traverse(root1) == expected1);

    TreeNode* root2 = new TreeNode(9);
    std::vector<vector<int>> expected2 = {{9}};
    assert(sol.traverse(root2) == expected2);

    TreeNode* root3 = new TreeNode(1);
    root3->left = new TreeNode(2);
    root3->right = new TreeNode(3);
    root3->left->left = new TreeNode(4);
    root3->left->right = new TreeNode(5);
    root3->right->left = new TreeNode(6);
    root3->right->right = new TreeNode(7);
    std::vector<vector<int>> expected3 = {{1}, {3, 2}, {4, 5, 6, 7}};
    assert(sol.traverse(root3) == expected3);

    TreeNode* root4 = new TreeNode(1);
    root4->left = new TreeNode(2);
    root4->right = new TreeNode(3);
    root4->left->left = new TreeNode(4);
    root4->right->left = new TreeNode(5);
    root4->right->right = new TreeNode(6);
    root4->left->left->left = new TreeNode(7);
    std::vector<vector<int>> expected4 = {{1}, {3, 2}, {4, 5, 6}, {7}};
    assert(sol.traverse(root4) == expected4);

    cleanupTree(root1);
    cleanupTree(root2);
    cleanupTree(root3);
    cleanupTree(root4);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}

File: gtci/10-level-order-traversal/python3/5-zigzag-traversal.py
from collections import deque


class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None


class Solution:
    def traverse(self, root):
        if root is None:
            return []

        traversal = []

        queue = deque()
        queue.append(root)

        zigzag = False
        while queue:
            level_size = len(queue)

            level = deque()
            for _ in range(level_size):
                node = queue.popleft()

                if zigzag:
                    level.appendleft(node.val)
                else:
                    level.append(node.val)

                if node.left is not None:
                    queue.append(node.left)

                if node.right is not None:
                    queue.append(node.right)

            traversal.append(list(level))
            zigzag = not zigzag

        return traversal


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(12)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(9)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(5)
    assert sol.traverse(root) == [[12], [1, 7], [9, 10, 5]]

    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    assert sol.traverse(root) == [[3], [20, 9], [15, 7]]

    print("All test cases passed.")
